### PR TITLE
Feature/point on link

### PR DESCRIPTION
**Pull request by dishadas168**
- PointOnLinks struct has replaced ContactPoints struct in Phase and WalkCycle classes. 

Now targets new PR #264

### CODE REVIEW

File: gtdynamics/utils/Phase.cpp [hunk 1]
```diff
@@ -26,7 +26,7 @@ using std::string;
 std::ostream &operator<<(std::ostream &os, const Phase &phase) {
   os << "[";
   for (auto &&cp : phase.contactPoints()) {
-    os << cp.first << ": " << cp.second << ", ";
+    os << cp.link->name() << ": " << cp.point << ", ";
```
Comment (dellaert) [inline gtdynamics/utils/Phase.cpp:29]: @jerredchen Try fixing test by changing to point.transpose() ? Feel free to directly add this branch now.

Comment (jerredchen) [reply gtdynamics/utils/Phase.cpp:29]: Sounds good, will do.

Comment (jerredchen) [reply gtdynamics/utils/Phase.cpp]: @dellaert Just updated it, and it's passing the unit test now. Thanks!

**Review by dellaert — CHANGES_REQUESTED** [review]
Lots of changes needed - maybe structural ones (move to cpp, call children methods) should be in separate PR, with this one restricted to the PointOnLink change. Functional constructor changes should be done here.

File: gtdynamics.i [hunk 2]
```diff
@@ -826,16 +834,16 @@ class WalkCycle {
...
+  initContactPointGoal(const gtdynamics::Robot& robot, 
+                       const gtdynamics::ContactAdjustments &contact_adjustments) const;
   std::vector<string> swingLinks(size_t p) const;
   gtsam::NonlinearFactorGraph
   contactPointObjectives(const gtsam::Point3 &step,
-                         const gtsam::SharedNoiseModel &cost_model,
-                         const gtdynamics::Robot &robot, size_t k_start,
-                         std::map<string, gtsam::Point3> @cp_goals) const;
+                         const gtsam::SharedNoiseModel &cost_model, size_t k_start,
+                         gtdynamics::ContactGoals @cp_goals) const;
```
[the remark this answers is not in the document]
Comment (dellaert) [reply gtdynamics.i:846]: Issue #254

File: tests/testPhase.cpp [hunk 3]
```diff
@@ -30,46 +30,46 @@ TEST(Phase, All) {
...
   EXPECT_LONGS_EQUAL(num_time_steps * 5, factors.size());
-  EXPECT(assert_equal(goal, cp_goals["tarsus_2_L2"]));
-  EXPECT(assert_equal(goal, cp_goals["tarsus_1_L1"]));
-  EXPECT(assert_equal(goal, cp_goals["tarsus_3_L3"]));
-  EXPECT(assert_equal(Point3(goal + step), cp_goals["tarsus_4_L4"]));
-  EXPECT(assert_equal(Point3(goal + step), cp_goals["tarsus_5_R4"]));
+  EXPECT(assert_equal(goal, cp_goals.at(0).goal_point));
+  EXPECT(assert_equal(goal, cp_goals.at(1).goal_point));
+  EXPECT(assert_equal(goal, cp_goals.at(2).goal_point));
+  EXPECT(assert_equal(Point3(goal + step), cp_goals.at(3).goal_point));
```
[the remark this answers is not in the document]
Comment (dellaert) [reply tests/testPhase.cpp:71]: Issue #254

File: examples/example_spider_walking/main.cpp [hunk 4]
```diff
@@ -39,39 +39,38 @@ using gtsam::noiseModel::Unit;
...
-  Phase stationary(40);
-  stationary.addContactPoints(links, contact_in_com);
-
-  Phase odd(20);
-  odd.addContactPoints(
-      {{"tarsus_1_L1", "tarsus_3_L3", "tarsus_5_R4", "tarsus_7_R2"}},
-      contact_in_com);
+Trajectory getTrajectory(const Robot& robot, size_t repeat) {
+  // TODO(frank): copy/paste from testSpiderWalking :-(
+  vector<LinkSharedPtr> odd_links = {
```
[the remark this answers is not in the document]
Comment (dellaert) [reply examples/example_spider_walking/main.cpp:44]: Done

File: gtdynamics/utils/Phase.h [hunk 5]
```diff
@@ -35,56 +36,64 @@ class Phase {
...
+   * @param[in] contact_in_com Point of contact on link.
    */
-  void addContactPoint(const std::string &link_name,
-                       const gtsam::Point3 &point) {
-    // Check if link exists in the robot
-    auto ret = contact_points_.emplace(link_name, ContactPoint{point, 0});
-    if (!ret.second) {
-      throw std::runtime_error("Multiple contact points for link " + link_name);
-    }
+  void addContactPoint(const LinkSharedPtr &link,
```
[the remark this answers is not in the document]
Comment (dellaert) [reply gtdynamics/utils/Phase.h:54]: Done

**Comment by dellaert** [comment]
BTW @jerredchen i’m addressing this. Don’t touch for now :-)

**Comment by jerredchen** [comment]
@dellaert Okay sounds good! I'll still be keeping tabs on the thread in case there's anything you'd like me to update.

File: gtdynamics/utils/Trajectory.h [hunk 6]
```diff
@@ -52,15 +52,22 @@ class Trajectory {
...
    *
-   * @param robot       Robot configuration.
    * @param walk_cycle  The Walk Cycle for the robot.
    * @param repeat      The number of repetitions for each phase of the gait.
    */
   Trajectory(const WalkCycle &walk_cycle, size_t repeat)
-      : repeat_(repeat), walk_cycle_(walk_cycle) {}
+      : repeat_(repeat), walk_cycle_(walk_cycle) {};
+  
+  ContactPoints toContactPointsObject(const PointOnLinks &contact_points) const{
```
[the remark this answers is not in the document]
Comment (dellaert) [reply gtdynamics/utils/Trajectory.h]: Obsolete

File: gtdynamics/utils/Trajectory.h [hunk 7]
```diff
@@ -52,15 +52,22 @@ class Trajectory {
...
+    ContactPoints cps;
+    for(auto&& cp : contact_points){
+      cps.emplace(cp.link->name(), ContactPoint{cp.point, 0});
+    }
+    return cps;
+  }
 
   /**
-   * @fn Returns a vector of ContactPoints objects for all phases after
+   * @fn Returns a vector of PointOnLinks objects for all phases after
```
[the remark this answers is not in the document]
Comment (dellaert) [reply gtdynamics/utils/Trajectory.h:64]: Will be refactored

File: gtdynamics/utils/WalkCycle.cpp [hunk 8]
```diff
@@ -20,6 +20,34 @@ using gtsam::NonlinearFactorGraph;
 using gtsam::Point3;
 using std::string;
 
+void WalkCycle::addPhase(const Phase &phase) {
```
[the remark this answers is not in the document]
Comment (dellaert) [reply gtdynamics/utils/WalkCycle.cpp:23]: Not now

File: gtdynamics/utils/WalkCycle.cpp [hunk 9]
```diff
@@ -33,21 +61,32 @@ void WalkCycle::print(const string &s) const {
...
+bool hasPoint(const ContactGoals& cp_goals, const PointOnLink &contact_point) {
+    auto it = std::find_if(
+        cp_goals.begin(), cp_goals.end(),
+        [&](const ContactGoal &cp_goal){
+          return cp_goal.point_on_link.link->name() == contact_point.link->name();
+        });
+    return !(it == cp_goals.end());
+}
+
+ContactGoals WalkCycle::initContactPointGoal(
```
[the remark this answers is not in the document]
Comment (dellaert) [reply gtdynamics/utils/WalkCycle.cpp:73]: Will be refactored

File: gtdynamics/utils/WalkCycle.h [hunk 10]
```diff
@@ -93,8 +97,9 @@ class WalkCycle {
    * @fn Returns the initial contact point goal for every contact link.
    * @return Map from link name to goal points.
    */
-  std::map<std::string, gtsam::Point3> initContactPointGoal(
-      const Robot& robot, double ground_height = 0) const;
+  ContactGoals initContactPointGoal(
+      const Robot &robot,
+      const ContactAdjustments &contact_adjustments = {}) const;
```
[the remark this answers is not in the document]
Comment (dellaert) [reply gtdynamics/utils/WalkCycle.h:102]: Issue created

File: tests/testKinematicsPhase.cpp [hunk 11]
```diff
@@ -25,6 +25,13 @@ using std::string;
 #include "contactGoalsExample.h"
 
 TEST(Phase, InverseKinematics) {
+  // Load robot and establish contact/goal pairs
+  using namespace contact_goals_example;
+
+  constexpr size_t num_time_steps = 5;
+  Phase phase0(num_time_steps);
+  phase0.addContactPoint(LH, contact_in_com);
```
[the remark this answers is not in the document]
Comment (dellaert) [reply tests/testKinematicsPhase.cpp]: Done

File: gtdynamics/utils/Trajectory.h [hunk 12]
```diff
@@ -52,15 +52,22 @@ class Trajectory {
...
    * @param repeat      The number of repetitions for each phase of the gait.
    */
   Trajectory(const WalkCycle &walk_cycle, size_t repeat)
-      : repeat_(repeat), walk_cycle_(walk_cycle) {}
+      : repeat_(repeat), walk_cycle_(walk_cycle) {};
+  
+  ContactPoints toContactPointsObject(const PointOnLinks &contact_points) const{
+    ContactPoints cps;
+    for(auto&& cp : contact_points){
+      cps.emplace(cp.link->name(), ContactPoint{cp.point, 0});
```
[the remark this answers is not in the document]
Comment (dellaert) [reply gtdynamics/utils/Trajectory.h]: obsolete

File: gtdynamics/utils/Trajectory.h [hunk 13]
```diff
@@ -88,11 +95,11 @@ class Trajectory {
     ContactPoints phase_2_cps;
 
     for (size_t p = 0; p < walk_cycle_.numPhases(); p++) {
-      phase_1_cps = phases[p].contactPoints();
+      phase_1_cps = toContactPointsObject(phases[p].contactPoints());
```
[the remark this answers is not in the document]
Comment (dellaert) [reply gtdynamics/utils/Trajectory.h]: removed

File: gtdynamics/utils/WalkCycle.cpp [hunk 14]
```diff
@@ -33,21 +61,32 @@ void WalkCycle::print(const string &s) const {
   std::cout << (s.empty() ? s : s + " ") << *this << std::endl;
 }
 
-std::map<string, Point3> WalkCycle::initContactPointGoal(
-    const Robot &robot, double ground_height) const {
-  std::map<string, Point3> cp_goals;
+bool hasPoint(const ContactGoals& cp_goals, const PointOnLink &contact_point) {
```
[the remark this answers is not in the document]
Comment (dellaert) [reply gtdynamics/utils/WalkCycle.cpp:64]: done

File: gtdynamics/utils/WalkCycle.h [hunk 15]
```diff
@@ -23,14 +23,32 @@
...
+
 /**
  * @class WalkCycle class stores the sequence of phases
  * in a walk cycle.
  */
 class WalkCycle {
  protected:
   std::vector<Phase> phases_;     ///< Phases in walk cycle
-  ContactPoints contact_points_;  ///< All contact points
+  PointOnLinks contact_points_;  ///< All contact points
```
[the remark this answers is not in the document]
Comment (dellaert) [reply gtdynamics/utils/WalkCycle.h:51]: Will be refactored

File: gtdynamics/utils/WalkCycle.h [hunk 16]
```diff
@@ -73,14 +81,10 @@ class WalkCycle {
...
-  size_t numTimeSteps() const {
-    size_t num_time_steps = 0;
-    for (const Phase& p : phases_) num_time_steps += p.numTimeSteps();
-    return num_time_steps;
-  }
+  size_t numTimeSteps() const;
 
   /// Return all the contact points.
-  const ContactPoints& contactPoints() const { return contact_points_; }
+  const PointOnLinks& contactPoints() const { return contact_points_; }
```
[the remark this answers is not in the document]
Comment (dellaert) [reply gtdynamics/utils/WalkCycle.h:87]: will be refactored

File: python/notebooks/spider.ipynb [hunk 17]
```diff
@@ -112,7 +112,8 @@
     "graph = trajectory.multiPhaseFactorGraph(graph_builder, collocation, mu)\n",
     "\n",
     "# Build the objective factors.\n",
-    "ground_height = 1.0 \n",
+    "ground_height = []\n",
+    "ground_height.append(gtd.ContactAdjustment(\"body\", np.array([0,0,-1.0])))\n",
```
[the remark this answers is not in the document]
Comment (dellaert) [reply python/notebooks/spider.ipynb:116]: Issue created

**Review by dellaert — APPROVED** [review]
Looks good now after substantive changes from myself, should have a review pass by, say, @jerredchen

**Pull request merged**